feat: introducing `self.call_self` and `self.enqueue_self`

benesjan opened the pull request:
Fixes F-23

In this PR I add 4 new structs to the `ContractSelf` struct (`CallSelf`, `CallSelfStatic`, `EnqueueSelf` and `EnqueueSelfStatic`) that allow for the following API when calling functions on the same contract:

```rust
   self.call_self.some_private_function(args)
   self.call_self_static.some_view_function(args)
   self.enqueue_self.some_public_function(args)
   self.enqueue_self_static.some_view_function(args)
```

benesjan commented:
* **#18076** <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/18076?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/18076?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `next`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

benesjan commented on noir-projects/aztec-nr/aztec/src/macros/functions/call_interface_stubs.nr:
```diff
@@ -101,13 +121,13 @@ comptime fn create_private_stub(f: FunctionDefinition) -> Quoted {
 
 comptime fn create_private_static_stub(f: FunctionDefinition) -> Quoted {
-    let (fn_name, fn_parameters_list, serialized_args_slice_construction, serialized_args_array_name, fn_name_str, fn_name_len, fn_selector) =
+    let (fn_name, fn_parameters_list, serialized_args_array_construction, serialized_args_array_name, fn_name_str, fn_name_len, fn_selector) =
```
Unrelated but the naming here was stale (it's no a slice).

benesjan commented on noir-projects/aztec-nr/aztec/src/macros/functions/call_interface_stubs.nr:
```diff
@@ -176,3 +196,127 @@ comptime fn create_utility_stub(f: FunctionDefinition) -> Quoted {
+    }
+}
+
+// TODO(F-131): Drop the use of the CallInterface in the following 4 functions - it doesn't make sense to not not
+// perform the call directly using the context. I tried doing this already but it became a lot of pain due to the use of
+// slices and them being illegal to return from unconstrained functions. Makes sense to tackle this when cleaning up the
+// call interface code.
+// Note: Once we get rid of the structs we will be able to merge some of the static and non-static stub functions.
```
Would tackle this in the future refactor of the call interfaces.

benesjan commented on noir-projects/noir-contracts/contracts/docs/docs_example_contract/src/main.nr:
```diff
@@ -157,7 +157,20 @@ pub contract DocsExample {
             let mut context = PrivateContext::new(inputs, args_hash);
             let storage = Storage::init(&mut context);
-            aztec::contract_self::ContractSelf::new_private(&mut context, storage)
+            // The call_self, enqueue_self, call_self_static and enqueue_self_static are structs generated by the
+            // #[aztec] macro. It's impractical to show define them in this example, so we just use unit types here.
+            let call_self = ();
+            let enqueue_self = ();
+            let call_self_static = ();
+            let enqueue_self_static = ();
```
This contract feels stale and I doubt it's very useful anyway which I am using as justification for admittedly being lazy here.

nventuro replied:
Yeah. We have nargo expand now as well.

benesjan commented on noir-projects/aztec-nr/aztec/src/contract_self.nr:
```diff
@@ -133,16 +193,25 @@ impl<Storage> ContractSelf<&mut PrivateContext, Storage> {
     /// Creates a new `ContractSelf` instance for a public function.
     ///
     /// This constructor is called automatically by the macro system and should not be called directly.
-    ///
-    /// # Parameters
-    /// - `context`: A mutable reference to the public execution context
-    /// - `storage`: The contract's storage instance
```
We also didn't have the param desc in new_private and I think it's unnecessary here given that it's in the docs of ContractSelf.

nventuro commented on noir-projects/noir-contracts/contracts/test/avm_test_contract/src/main.nr:
We probably want to have a version of `call` and `call_self` where we can set the gas, e.g.
```noir
self.call_sef_opts(Options::gas(whatever)).foo();
```

Not now, but down the line it might make sense.

nventuro commented on noir-projects/aztec-nr/aztec/src/macros/functions/self_call_registry.nr:
```diff
@@ -0,0 +1,81 @@
+
+pub comptime fn get_all_self_call_stubs(
+    m: Module,
+) -> (Quoted, Quoted, Quoted, Quoted, Quoted, Quoted) {
```
Maybe put these in a struct so that they are named. Right now we rely on the order here being the same as in the callsite. Easy to make mistakes.

benesjan replied:
Addressed in https://github.com/AztecProtocol/aztec-packages/pull/18076/commits/3e4ffa1f21001e8714ca0e574f32b1cd1118042d

nventuro commented on noir-projects/aztec-nr/aztec/src/macros/functions/utils.nr:
```diff
@@ -229,7 +234,10 @@ pub(crate) comptime fn transform_public(f: FunctionDefinition) {
+            let self_address = context.this_address();
+            let call_self: CallSelf<aztec::context::public_context::PublicContext> = CallSelf { address: self_address, context };
+            let call_self_static: CallSelfStatic<aztec::context::public_context::PublicContext> = CallSelfStatic { address: self_address, context };
+            aztec::contract_self::ContractSelf::new_public(context, storage, call_self, call_self_static)
```
Does this mean that the private fns are available here to call?

benesjan replied:
It doesn't because the `PublicContext` is used as a generic in `CallSelf` and the implementation of `CallSelf` for `PublicContext` contains only the public methods (see [here](https://github.com/AztecProtocol/aztec-packages/blob/3e4ffa1f21001e8714ca0e574f32b1cd1118042d/noir-projects/aztec-nr/aztec/src/macros/aztec.nr#L322)).

Checked and the LSP manages to show this nicely:

<img width="853" height="574" alt="image" src="https://github.com/user-attachments/assets/07db6210-64b8-4ea0-84fb-f53ab8c39408" />

benesjan committed:
feat: introducing `self.call_self` and `self.enqueue_self`

Fixes F-23

In this PR I add 4 new structs to the `ContractSelf` struct (`CallSelf`, `CallSelfStatic`, `EnqueueSelf` and `EnqueueSelfStatic`) that allow for the following API when calling functions on the same contract:

```rust
   self.call_self.some_private_function(args)
   self.call_self_static.some_view_function(args)
   self.enqueue_self.some_public_function(args)
   self.enqueue_self_static.some_view_function(args)
```
